modules/rotating-certificate-authority.adoc

```diff
@@ -0,0 +1,54 @@
+// Module included in the following assemblies:
+//
+// security/certificate_types_descriptions/etcd-certificates.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="rotating-certificate-authority_{context}"]
+== Rotating the etcd certificate
+
+Rotate the `etcd` certificate before it expires.
+
+.Procedure
+
+. Verify the remaining lifetime of the new signer certificate:
++
+[source,terminal]
+----
+$ oc get secret -n openshift-etcd etcd-signer -ojsonpath='{.metadata.annotations.auth\.openshift\.io/certificate-not-after}'
+----
+
+. Re-create the signer, if the remaining lifetime is close to the current date, by deleting the signer and wait for the static pod rollout:
++
+[source,terminal]
+----
+$ oc delete secret -n openshift-etcd etcd-signer
+----
++
+[source,terminal]
+----
+$ oc wait --for=condition=Progressing=False --timeout=15m clusteroperator/etcd
+----
+
+. After `etcd` restarts, you can switch the original certificate authority(CA) in the `openshift-config` namespace with the new, rotated one in `openshift-etcd`.
++
+[source,terminal]
+----
+$ oc get secret etcd-signer -n openshift-etcd -ojson | jq 'del(.metadata["namespace","creationTimestamp","resourceVersion","selfLink","uid"])' | oc apply -n openshift-config -f -
+----
++
+[source,terminal]
+----
+$ oc wait --for=condition=Progressing=False --timeout=25m clusteroperator/etcd
+----
++
+[source,terminal]
+----
+$ oc wait --for=condition=Progressing=False --timeout=25m clusteroperator/kube-apiserver
+----
+
+You can also use this single command to switch the CA:
+
+[source,terminal]
+----
+$ oc adm wait-for-stable-cluster
+----
```

Comment:
We should document a single way to do this, unless there is a specific use case for a second method. In this instance, if the two methods are equivalent, I would lean towards the single-command step for simplicity.

```suggestion
. After `etcd` restarts, switch the original CA in the `openshift-config` namespace with the new, rotated one in `openshift-etcd` by running the following command:
+
[source,terminal]
----
$ oc adm wait-for-stable-cluster
----
```

If you need to use the multi-command version, it should be split into substeps (similar to my suggestion for step 2 above) so we have a single command per step.

Reply:
The `oc adm wait-for-stable-cluster` cmd does the equivalent check as the previous two cmds (and more) but will wait until that required conditions are true for a period of 5 mins (`--minimum-stable-period 5m`) so in practice it might take longer to wait on that step than the individual cmds.
```console
$ oc adm wait-for-stable-cluster -h
Wait for all OCP v4 clusteroperators to report Available=true, Progressing=false, Degraded=false.

Examples:
  # Wait for all clusteroperators to become stable
  oc adm clusteroperator wait-for-stable-cluster

  # Consider operators to be stable if they report as such for 5 minutes straight
  oc adm clusteroperator wait-for-stable-cluster --minimum-stable-period 5m

Options:
    --minimum-stable-period=5m0s:
	minimum duration to consider a cluster stable. Defaults to 5 minutes.

    --timeout=1h0m0s:
	duration before the command times out. Defaults to 1 hour.
```

That may be a long time from the user's perspective so maybe a note here that the `oc adm wait-for-stable-cluster` will wait for a minimum of 5 mins by default.

@tjungblu Do you think we need to wait that long in practice or if we can shorten it e.g `oc adm wait-for-stable-cluster --minimum-stable-period 2m`

Reply:
2m is enough for sure

Reply:
Sounds good, thanks for confirming. Then we can combine those into the following. Note we still need the step to switch out the CA first and then wait with the singular cmd for 2m.
```suggestion
. After `etcd` restarts, you can switch the original certificate authority(CA) in the `openshift-config` namespace with the new, rotated one in `openshift-etcd`.
+
[source,terminal]
----
$ oc get secret etcd-signer -n openshift-etcd -ojson | jq 'del(.metadata["namespace","creationTimestamp","resourceVersion","selfLink","uid"])' | oc apply -n openshift-config -f -
----
+

You can then wait for the cluster operators to rollout and become stable:

[source,terminal]
----
$ oc adm wait-for-stable-cluster --minimum-stable-period 2m
----
```